### ISSUE TITLE
support sharing host fonts with toolbox

### ISSUE THREAD

**Issue by juhp** [containers/toolbox#1594]
**Is your feature request related to a problem? Please describe.**

It would be cool to have a 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
Add any other context or screenshots about the feature request here. If a different tool has the functionality you're requesting, share it here.

**Comment by debarshiray** [comment]
> It would be cool to have a consistent way to use host fonts from within toolbox.

Yes, indeed.  I don't know enough about fonts, so I never tried this on my own.

> **Describe the solution you'd like**
> 
> I may be tricky to find a general consistent approach, though what I personally do is for Fedora toolboxes (on Fedora):
> 
> `sudo ln -s /run/host/usr/share/fonts /usr/local/share/`
> 
> without running fc-cache.

Is it noticeably slower if the font information cache is absent?

Where does the cache live?  The `fc-cache(1)` manual says `%cachedir%/*-%arch%.cache-%version%` but I don't know what the values of those variables are.  I don't see anything in `/var/cache`, but `~/.cache/fontconfig/` has some files.

Are the formats of the font files and the cache stable enough to be mixed and matched across the host and containers?

If they are stable enough, maybe we can bind mount `/run/host/usr/share/fonts` to the container's `/usr/share/fonts`, and add `/usr/share/fonts` to `%_netsharedpath` in `/usr/lib/rpm/macros.d/macros.toolbox` so that RPMs inside the container leave it alone?  This way we could have less content in the Toolbx images, which is always better.

What do you think?

> Though apparently bad things can happen with this for Silverblue with fc-cache due to the timestamps.

> FWIW, distrobox-init seems to do something like this: https://github.com/89luca89/distrobox/blob/18053c254a83750c49c08e58df2e48a0f04aef48/distrobox-init#L2007

They are doing a bind mount instead of a symbolic link, but it's the same thing for the purposes of this issue.

**Comment by juhp** [comment]
> Is it noticeably slower if the font information cache is absent?

Not to me. There may be a small start-up delay for GUI apps the first time?

The other discrepancy is missing /etc/fonts/conf.d/ files from the host.
But at a basic level just linking the fonts directory seems to work reasonably okay on Silverblue in my experience: so I don't feel it is harmful. Though it is possible if there are many extra fonts in installed on the host, the results might not be as predictable: I think in some sense it dependencies on the amount of overlapping font coverage - the default fontconfig conf.d rules should provide some minimal order at least.
(I think distrobox also doesn't link conf.d)

----

I believe the default system cache is in `/usr/lib/fontconfig/cache/`.

I think potential problems with sharing the fontconfig cache include:

- it is possible, specially for older OSes, for the host and toolbox fontconfig's to be incompatible (we have seen this also for flatpaks in the recent past, though I believe the handling has improved for recent fontconfig versions)
- also naively it may interfere with any fonts packages installed in the toolbox

In theory something could be done for these perhaps - though it may require non-trivial changes on the fontconfig side.

Anyway @tagoh is the expert here, I am just sharing my limited experience on this.

**Comment by debarshiray** [comment]
Thanks for bringing up Flatpak.  It is always a good data point.  How do fonts work in Flatpaks?  Do the Flatpak containers have their own fonts, presumably through the runtime?  Or are the fonts shared with the host?

I didn't know about `/etc/fonts/conf.d`.  I see that it has symbolic links to configuration files installed by the fonts packages in `/usr/share/fontconfig`.

I think I don't know how fonts work and that's preventing me from imagining all the possibilities.

Maybe we can work out the use-cases around fonts?  I can imagine that most users who are not font developers will be happy to have the fonts shared with the host.  What about font developers?  Will it somehow get in their way?

If we do bind mount the host's `/usr/share/fonts` into the container's `/usr/share/fonts`, then setting `%_netsharedpath` (see above) should block RPM from adding files to that location inside the container.

**Comment by debarshiray** [comment]
Ping @juhp @tagoh

**Comment by tagoh** [comment]
I thought I mentioned something about fontconfig and flatpak somewhere else. let me add a few comments again.

> Thanks for bringing up Flatpak. It is always a good data point. How do fonts work in Flatpaks? Do the Flatpak containers have their own fonts, presumably through the runtime? Or are the fonts shared with the host?

Both. Because of this, Flatpak has a special configuration for fontconfig to remap host font directory to another and make it possible to reuse host cache files because cache files contains a font path where points to a font file and it is based on directory structure at host. So without this, Flatpaks can't find a font out against it.

> If we do bind mount the host's `/usr/share/fonts` into the container's `/usr/share/fonts`, then setting `%_netsharedpath` (see above) should block RPM from adding files to that location inside the container.

So an idea like the quoted your mention could be brought up to address it maybe. However I don't think bind-mounting /usr/share/fonts into /usr/share/fonts works. That will be a cause preventing to install a container-specific font and that is why Flatpak do remap a font directory.

However, everything doesn't perfectly works even with it. Font packages more or less has own config files which has an important information to decide a generic family, such as sans-serif, serif, and monospace etc.   We can't get the best font effectively without it because there are no way to know what is a generic-family name for a font "Foo".

Other than that, there are an issue that is reproducible behavior against config. A font preference will be different against the order of config. The best font behavior could be different compared to host because a container can have own font packages installed.  There might be other issues I'm not aware of.  Anyway, Flatpak doesn't share config files from host at the moment.

**Comment by debarshiray** [comment]
Thanks for the explanation, @tagoh !

I found `flatpak_ensure_data_dir` and `add_font_path_args` in Flatpak's [common/flatpak-run.c](https://github.com/flatpak/flatpak/blob/main/common/flatpak-run.c) based on your description.  I will try to read through it and see if I can come up with something.

Is there a good way to test that the set-up works?  I don't get any hits for tests with `git grep font` in the Flatpak code.

**Comment by tagoh** [comment]
Yes, fontconfig has a test case for that. See https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blame/main/test/run-test.sh?ref_type=heads#L242 and https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blame/main/src/fccache.c?ref_type=heads#L161
That may looks complicated. The key point is:
```
<remap-dir as-path="/path/to/fontdir/at/host">/path/to/fontdir/at/sandbox</remap-dir>
```
This line basically do all the job. but please note that `<dir>` is still functional and if a font dir at sandbox has same font installed and set it like `<dir>/path/to/fontdir/somewhere/in/sandbox</dir>`, fontconfig may **update** cache with same md5, because fontconfig estimate md5sum for cache from font path (i.e. both /usr/share/fonts at host and /usr/share/fonts at sandbox has same md5sum no matter what they actually have). So we have another way to make it separate.
```
<dir salt="something">/path/to/fontdir/at/host/but/actually/in/sandbox</dir>
```
This allow us to create a cache with different md5sum. Also you could put it under a cache dir for sandbox only if you add dedicated `<cachedir>` for sandbox in prior to bind-mounted cache dir from host.

Please let me know if you need any further helps.

**Comment by debarshiray** [comment]
Thanks again, @tagoh !  I will let you know if I need more help.  :)
